File: Segundo/PDOO/Tema 1/Civitas/src/main/java/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual () {
    in = new Scanner (System.in);
  }
  
  public void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  public void pausa() {
    System.out.println ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
        System.out.println (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    System.out.println(separador);
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  public SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  public Respuestas comprar() {
    int opcion = menu ("Desea comprar la propiedad",
      new ArrayList<> (Arrays.asList("No","Si")));
    return (Respuestas.values()[opcion]);
  }

  public void gestionar () {
      int nGestion = menu ("Seleccione la operación a realizar",
      new ArrayList<> (Arrays.asList("Vender","Hipotecar","Cancelar hipoteca","Construir casa","Construir hotel","Terminar")));
    iGestion = nGestion;
    if (iGestion != 5) {
        ArrayList<TituloPropiedad> propiedades = juegoModel.getJugador().getPropiedades();
        ArrayList<String> nombres = new ArrayList<>(Arrays.asList());
        for (int i = 0; i < propiedades.size(); i++) {
        nombres.add(propiedades.get(i).getNombre());
        }
        int nPropiedad = menu ("Seleccione la propiedad sobre la que desea realizar la gestión", nombres);
        iPropiedad = nPropiedad;
    }
  }
  
  public int getGestion(){
    return iGestion;
  }
  
  public int getPropiedad(){
    return iPropiedad;
  }
    

  public void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      System.out.println(separador);
      System.out.println("La siguiente operacion a realizar es " + operacion.toString());
  }


  public void mostrarEventos() {
      if (Diario.getInstance().eventosPendientes())
        System.out.println(separador);
      while (Diario.getInstance().eventosPendientes()) {
          System.out.println(Diario.getInstance().leerEvento());
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();

    }
  
  public void actualizarVista(){
      System.out.println(separador);
      System.out.println(juegoModel.getJugador().toString());
      System.out.println(juegoModel.getCasillaActual().toString());
  }
  
  public void ranking() {
      System.out.println(separador);
      ArrayList<Jugador> ranking = juegoModel.ranking();
        for (int i = 1; i <= ranking.size(); i++) {
            System.out.println(ranking.get(i-1).toString() + " Además ha quedado en posición " + i + ".");
        }
  }
}

File: Segundo/PDOO/CivitasRuby/VistaTextual/vista_textual.rb
#encoding:utf-8
require_relative '../Civitas/CivitasJuego'
require_relative '../Civitas/Operaciones_juego'
require_relative '../Civitas/Respuestas'
require_relative '../Civitas/SalidasCarcel'
require 'io/console'

module Civitas

  class Vista_textual


    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    
    def comprar
      lista_Respuestas = [Respuestas::SI, Respuestas::NO]
      opciones = ["Si", "No"]
      opcion = menu("Desea comprar la propiedad?", opciones)
      return lista_Respuestas[opcion]
    end

    def gestionar
      opciones = ["Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar"]
      opcion = menu("Seleccione la gestión a realizar", opciones)
      @iGestion = opcion
      if opcion != 5
        propiedades = @juegoModel.getJugador().propiedades
        nombres = []
        propiedades.each { |p|
          nombres << p.nombre
        }
        ip = menu("Seleccione la propiedad en la que desea realizar la gestión", nombres)
        @iPropiedad = ip
      end
    end

    def getGestion
      return @iGestion
    end

    def getPropiedad
      return @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      print "\nLa siguiente operación a realizar es " + operacion.to_s + ".\n\n"
    end

    def mostrarEventos
      while(Diario.instance.eventos_pendientes())
        puts Diario.instance.leer_evento()
      end
    end

    def salirCarcel
      lista_SalidasCarcel = [SalidasCarcel::PAGANDO, SalidasCarcel::TIRANDO]
      opciones =["Pagando", "Tirando"]
      opcion = menu("Seleccione como desea salir de la cárcel", opciones)
      return lista_SalidasCarcel[opcion]
    end

    def setCivitasJuego(civitas)
      @iGestion
      @iPropiedad  
      @juegoModel=civitas
      self.actualizarVista
    end

    def actualizarVista
      jugador = @juegoModel.getJugador()
      print "\nEl jugador actual es " + jugador.nombre + " y tiene un saldo de " + jugador.saldo.to_s + ".\n"
      casilla = @juegoModel.getCasillaActual()
      print "La casilla actual es " + casilla.nombre + " y es de tipo " + casilla.toString() + "."
      if (casilla.is_a?(CasillaCalle))
        if (!casilla.titulo.tienePropietario)
          print " Además, tiene un precio de " + casilla.cantidad.to_s + "."
        end
      end
      print "\n"
    end

    def ranking
      jugadores = @juegoModel.ranking()
      posicion = 1

      jugadores.each { |j|
        puts "El jugador " + j.nombre + " ha quedado en la posición " + posicion.to_s + " con un saldo de " + j.saldo.to_s + ".\n"
        posicion = posicion + 1
      }
    end

    
  end

end
